lib/compiler/src/graph_utils.cc

```diff
@@ -0,0 +1,103 @@
+#include "graph_utils.h"
+
+namespace FlexFlow {
+
+SerialParallelDecomposition
+    get_serial_parallel_decomposition(OptimizerPCG const &pcg) {
+  return get_serial_parallel_decomposition(
+      unsafe_view_as_digraph(MultiDiGraphView(pcg)));
+}
+
+std::vector<MultiDiEdge> get_sorted_node_input_edges(OptimizerPCG const &pcg,
+                                                     Node const &n) {
+  std::unordered_map<std::size_t, std::unordered_set<MultiDiEdge>>
+      incoming_edges = get_incoming_edges_by_idx(MultiDiGraphView(pcg), n);
+
+  std::vector<MultiDiEdge> result;
+  for (std::size_t i = 0; i < incoming_edges.size(); i++) {
+    result.push_back(get_only(incoming_edges.at(i)));
+  }
+
+  return result;
+}
+
+std::unordered_map<MultiDiEdge, ParallelTensorShape>
+    infer_tensor_shapes(OptimizerPCG const &pcg) {
+  std::unordered_map<MultiDiEdge, ParallelTensorShape> result;
+  for (Node const &n : get_topological_ordering(MultiDiGraphView(pcg))) {
+    PCGOperatorAttrs op = pcg.at(n);
+
+    std::vector<ParallelTensorShape> input_tensor_shapes =
+        vector_transform([&](MultiDiEdge const &e) { return result.at(e); },
+                         get_sorted_node_input_edges(pcg, n));
+
+    std::vector<ParallelTensorShape> output_tensor_shapes =
+        get_output_shapes(op, input_tensor_shapes);
+
+    auto outgoing_edges = get_outgoing_edges_by_idx(MultiDiGraphView(pcg), n);
+
+    for (std::size_t i = 0; i < output_tensor_shapes.size(); i++) {
+      if (contains_key(outgoing_edges, i)) {
+        for (MultiDiEdge const &e : outgoing_edges.at(i)) {
+          result.insert({e, output_tensor_shapes[i]});
+        }
+      }
+    }
+  }
+
+  assert(result.size() == get_edges(MultiDiGraphView(pcg)).size());
+
+  return result;
+}
+
+template <typename NodeLabel,
+          typename EdgeLabel,
+          typename InputLabel = EdgeLabel,
+          typename OutputLabel = EdgeLabel>
+LabelledOpenMultiDiGraph<NodeLabel, EdgeLabel, InputLabel, OutputLabel>
+    get_subgraph(LabelledOpenMultiDiGraph<NodeLabel,
```

Comment:
Is this best done this way or by first converting a `LabelledOpen` to a `LabelledDownwardOpen` or `LabelledUpwardOpen` and then using `get_subgraph`?

Reply:
I just wonder if the second would be a bit cleaner and more orthogonal